src/rgw/rgw_rest_s3.cc

```diff
@@ -4576,6 +4575,7 @@ RGWOp *RGWHandler_REST_Obj_S3::op_delete()
     return new RGWAbortMultipart_ObjStore_S3;
 }
 
+RGWOp* create_s3select_op();
```

Comment:
the decl here is strange;  I read ahead, and I now know this is provided in rgw_s3select.cc--probably should put this decl in rgw_s3select.h?  also suggested below, put it in a namespace

Reply:
`RGWOp* create_s3select_op();` is the only thing which link RGW to s3select operation,
it could reside on `rgw_rest_s3.h`

the RGW is not aware of s3select internal definitions.

`rgw_s3select.h` is exposed only to the implementation layer, i.e. `rgw_s3select.cc`

Reply:
This doesn't really make sense.  If the "module" is rgw_select.cc, and provides create_s3select_op(), it's part of the module interface;  it's the module interface that should provide the decl, not rgw_rest_s3.h, which is part of the consuming module

Reply:
yes, `RGWOp* create_s3select_op();` is part of the "module".
most of `rgw_s3select.h` contain non-relevant definitions to RGW, this should be hidden.
`RGWOp* create_s3select_op();` is the only relevant definition.

Reply:
if you really need a private header for rgw_s3elect.cc, you could create something like rgw_select_private.h

Reply:
(which would include rgw_s3select.h)

Reply:
:+1: